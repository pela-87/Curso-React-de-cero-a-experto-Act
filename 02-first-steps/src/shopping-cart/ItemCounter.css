


.item-row {
    display: flex;
    align-Items: center;
    gap: 10px;
    margin-Top: 10px;
}

.item-text {
    width: 150px;
}