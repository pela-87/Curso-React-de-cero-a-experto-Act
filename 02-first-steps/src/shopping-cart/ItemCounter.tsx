import { useState } from "react";


//import './ItemCounter.css';
//Utilizando CSS como module
import style from './ItemCounter.module.css';

interface Props { 
    name : string,
    quantity?: number,
};

export const ItemCounter = ( {name, quantity = 1} : Props) => {

  const [count, setCount] = useState(quantity);

  const handleAdd = () => {
    setCount(count + 1);
  }
  const handleSubtract = () => {
    if (count === 1) return;
    setCount(count - 1);
  }


  return (
    <section 
      //Dandole stylo como modulo.
      className= {style['item-row']}

      //Utilizando los style de css normal.
      //className="item-row"

      //dandole stylo desde codigo.
      // style={{
      //     display: 'flex',
      //     alignItems: 'center',
      //     gap: 10,
      //     marginTop: 10,
      // }}
    >
        <span            
            className={style['item-text']}
            style={{
              color: count === 1 ? 'red' : 'black',
            }}
        >
          {name} 
        </span>
        <button onClick={ handleAdd }>
          +1
        </button>
        <span>{count}</span>
        <button onClick={ handleSubtract }>
          -1
        </button>
    </section>
  )
}
