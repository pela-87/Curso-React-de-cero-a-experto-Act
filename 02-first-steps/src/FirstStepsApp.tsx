import { ItemCounter } from './shopping-cart/ItemCounter';

interface ItemInCart {
    productName: string,
    quantity: number,
}

const itemInCart: ItemInCart[] = [
    {productName: 'Nintendo Switch 2', quantity: 7},
    {productName: 'Dragon Ball 2', quantity: 4},
    {productName: 'Bomber-Man', quantity: 2},    
];

export function FirstStepsApp() {
    return (
        <>
            {/* <h2>Hola mundi.</h2>
            <p>Esto es un parrafo.</p>
            <button>Click me</button>

            <div>
                <h2>Hola dentro de un div</h2>
            </div> */}

            <h1>Carrito de compras</h1>

            {/* <ItemCounter name="Nintendo Swich 2" quantity={12} />
            <ItemCounter name="Dragon Ball 2" quantity={4}/>
            <ItemCounter name="Bomber-man 3" quantity={2}/> */}
            {
                itemInCart.map(({productName, quantity}) => (
                    <ItemCounter key={productName} name={productName} quantity={quantity}/>
                ))
            }
        </>
    )
}