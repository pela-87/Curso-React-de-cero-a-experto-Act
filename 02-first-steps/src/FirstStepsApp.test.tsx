import { afterEach, describe, expect, test, vi } from 'vitest';
import { FirstStepsApp } from './FirstStepsApp';
import { render, screen } from '@testing-library/react';
import { ItemCounter } from './shopping-cart/ItemCounter';

const mockItemCounter = vi.fn((props: unknown) => {
    return <div data-testid="ItemCounter" />;
});

vi.mock('./shopping-cart/ItemCounter', () => ({
    ItemCounter: (props: unknown) => mockItemCounter(props),
}));


// vi.mock('./shopping-cart/ItemCounter', () => ({
//     ItemCounter: (props: unknown) => <div data-testid="ItemCounter" name= { props.name} quantity= { props.quantity} />
// }));

describe('FirstStepApp', () => {

    //Esta linea limpia y prepara los mosck. Esto se ejecuta despues de cada prueba.
    afterEach(() => {
        vi.clearAllMocks();
    });


    test('Should match snapshot', () => {
        const { container } = render(<FirstStepsApp />);
        expect(container).toMatchSnapshot();        
    });
    
    test('should render the correct number of ItemCounter components', () => {
        render(<FirstStepsApp />);
        //screen.debug();
        const itemCointers = screen.getAllByTestId('ItemCounter');
        //console.log(itemCointers.length);
        expect(itemCointers.length).toBe(3);
    });

    test('should render ItemCounter with correct props', () => {
        render(<FirstStepsApp />);
        expect(mockItemCounter).toHaveBeenCalledTimes(3);
        expect(mockItemCounter).toHaveBeenCalledWith({
            name: 'Nintendo Switch 2', 
            quantity: 7,
        });
        expect(mockItemCounter).toHaveBeenCalledWith({
            name: 'Dragon Ball 2', 
            quantity: 4,
        });
        expect(mockItemCounter).toHaveBeenCalledWith({
            name: 'Bomber-Man', 
            quantity: 2,    
        });
    })
})