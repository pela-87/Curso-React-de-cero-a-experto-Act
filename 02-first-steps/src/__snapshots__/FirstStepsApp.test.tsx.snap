// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FirstStepApp > Should match snapshot 1`] = `
<div>
  <h1>
    Carrito de compras
  </h1>
  <div
    data-testid="ItemCounter"
  />
  <div
    data-testid="ItemCounter"
  />
  <div
    data-testid="ItemCounter"
  />
</div>
`;
